File: relationship-manager/recommendation-service/src/main/java/com/bank/rm/recommendation/service/RecommendationService.java
package com.bank.rm.recommendation.service;

import com.bank.rm.common.dto.RiskCategory;
import com.bank.rm.common.event.KafkaTopics;
import com.bank.rm.common.event.RecommendationGeneratedEvent;
import com.bank.rm.recommendation.dto.RecommendationDtos.*;
import dev.langchain4j.model.chat.ChatLanguageModel;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;
import java.util.*;

@Service
public class RecommendationService {
    private static final Logger log = LoggerFactory.getLogger(RecommendationService.class);
    private final ChatLanguageModel chatModel;
    private final KafkaTemplate<String, String> kafkaTemplate;
    private final ObjectMapper objectMapper;

    private static final List<SeedProduct> SEED_PRODUCTS = List.of(
        new SeedProduct("High-Yield Savings Account", "SAVINGS_ACCOUNT", 85, 15, Set.of(RiskCategory.CONSERVATIVE, RiskCategory.MODERATE)),
        new SeedProduct("Fixed Deposit - 5 Year", "FIXED_DEPOSIT", 90, 20, Set.of(RiskCategory.CONSERVATIVE)),
        new SeedProduct("Large Cap Equity Fund", "MUTUAL_FUND", 75, 25, Set.of(RiskCategory.MODERATE, RiskCategory.AGGRESSIVE)),
        new SeedProduct("ELSS Tax Saver Fund", "TAX_SAVER", 80, 10, Set.of(RiskCategory.MODERATE, RiskCategory.AGGRESSIVE)),
        new SeedProduct("National Pension System", "PENSION", 85, 10, Set.of(RiskCategory.CONSERVATIVE, RiskCategory.MODERATE)),
        new SeedProduct("Sovereign Gold Bond", "GOLD", 70, 10, Set.of(RiskCategory.CONSERVATIVE, RiskCategory.MODERATE)),
        new SeedProduct("PPF Account", "PPF", 90, 10, Set.of(RiskCategory.CONSERVATIVE)),
        new SeedProduct("Small Cap Equity Fund", "MUTUAL_FUND", 65, 20, Set.of(RiskCategory.AGGRESSIVE, RiskCategory.VERY_AGGRESSIVE)),
        new SeedProduct("Corporate Bond Fund", "BOND", 75, 15, Set.of(RiskCategory.MODERATE)),
        new SeedProduct("Multi Cap Growth Fund", "MUTUAL_FUND", 70, 30, Set.of(RiskCategory.AGGRESSIVE, RiskCategory.VERY_AGGRESSIVE)),
        new SeedProduct("Term Insurance Plan", "INSURANCE", 95, 5, Set.of(RiskCategory.CONSERVATIVE, RiskCategory.MODERATE, RiskCategory.AGGRESSIVE, RiskCategory.VERY_AGGRESSIVE))
    );

    public RecommendationService(ChatLanguageModel chatModel, KafkaTemplate<String, String> kafkaTemplate,
                                  ObjectMapper objectMapper) {
        this.chatModel = chatModel;
        this.kafkaTemplate = kafkaTemplate;
        this.objectMapper = objectMapper;
    }

    public RecommendationResponse generateRecommendations(UUID customerId, RecommendationRequest request) {
        List<ProductRecommendation> ruleBasedResults = new ArrayList<>();
        int rank = 1;
        for (SeedProduct sp : SEED_PRODUCTS) {
            if (sp.suitableFor().contains(request.riskCategory())) {
                ruleBasedResults.add(new ProductRecommendation(sp.name(), sp.category(),
                        sp.matchScore(), sp.allocationPercent(), null, rank++));
            }
        }

        List<ProductRecommendation> finalResults = aiRerank(ruleBasedResults, request);

        UUID recommendationId = UUID.randomUUID();
        publishEvent(new RecommendationGeneratedEvent(customerId, recommendationId, request.riskCategory()));

        return new RecommendationResponse(recommendationId, customerId, request.riskCategory(), finalResults);
    }

    private List<ProductRecommendation> aiRerank(List<ProductRecommendation> products, RecommendationRequest request) {
        try {
            StringBuilder productList = new StringBuilder();
            for (int i = 0; i < products.size(); i++) {
                productList.append(String.format("%d. %s (%s)\n", i, products.get(i).productName(), products.get(i).category()));
            }
            String prompt = String.format(
                "Given a %s risk profile customer, re-rank these products by suitability and " +
                "provide a one-line rationale for each:\n%s\n" +
                "Return as numbered list with format: rank. product_name | rationale",
                request.riskCategory(), productList);
            String response = chatModel.generate(prompt);

            List<ProductRecommendation> reranked = new ArrayList<>();
            int newRank = 1;
            for (ProductRecommendation p : products) {
                String rationale = extractRationale(response, p.productName());
                reranked.add(new ProductRecommendation(p.productName(), p.category(),
                        p.matchScore(), p.allocationPercent(),
                        rationale != null ? rationale : "Suitable for your risk profile", newRank++));
            }
            return reranked;
        } catch (Exception e) {
            log.warn("AI re-ranking failed, using rule-based order: {}", e.getMessage());
            return products;
        }
    }

    private String extractRationale(String response, String productName) {
        for (String line : response.split("\n")) {
            if (line.toLowerCase().contains(productName.toLowerCase().substring(0, Math.min(15, productName.length()))) && line.contains("|")) {
                return line.substring(line.indexOf("|") + 1).trim();
            }
        }
        return null;
    }

    private void publishEvent(RecommendationGeneratedEvent event) {
        try {
            kafkaTemplate.send(KafkaTopics.RECOMMENDATION_EVENTS, objectMapper.writeValueAsString(event));
        } catch (Exception e) { log.error("Failed to publish recommendation event: {}", e.getMessage()); }
    }

    private record SeedProduct(String name, String category, int matchScore, int allocationPercent, Set<RiskCategory> suitableFor) {}
}

File: relationship-manager/recommendation-service/src/main/kotlin/com/bank/rm/recommendation/service/RecommendationService.kt
package com.bank.rm.recommendation.service

import com.bank.rm.common.dto.RiskCategory
import com.bank.rm.common.event.*
import com.bank.rm.recommendation.domain.Recommendation
import com.bank.rm.recommendation.domain.RecommendedProduct
import com.bank.rm.recommendation.dto.*
import com.bank.rm.recommendation.repository.RecommendationRepository
import com.bank.rm.recommendation.repository.RecommendedProductRepository
import com.fasterxml.jackson.databind.ObjectMapper
import dev.langchain4j.model.chat.ChatLanguageModel
import org.slf4j.LoggerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Service
class RecommendationService(
    private val recommendationRepository: RecommendationRepository,
    private val recommendedProductRepository: RecommendedProductRepository,
    private val chatModel: ChatLanguageModel,
    private val kafkaTemplate: KafkaTemplate<String, DomainEvent>,
    private val objectMapper: ObjectMapper
) {
    private val log = LoggerFactory.getLogger(javaClass)

    @Transactional
    fun generateRecommendations(customerId: UUID, request: RecommendationRequest): RecommendationResponse {
        // Step 1: Rule-based product matching (base ranking)
        val baseProducts = matchProducts(request.riskCategory, request.customerProfile)

        // Step 2: AI re-ranking using LLM
        val rerankedProducts = try {
            reRankWithAI(baseProducts, request.customerProfile)
        } catch (e: Exception) {
            log.warn("AI re-ranking failed, using rule-based order: ${e.message}")
            baseProducts
        }

        // Step 3: Generate AI rationale for each recommendation
        val productsWithRationale = rerankedProducts.map { product ->
            val rationale = try {
                generateProductRationale(product, request.customerProfile)
            } catch (e: Exception) {
                "This product matches your ${request.riskCategory.name.lowercase()} risk profile."
            }
            product.copy(rationale = rationale)
        }

        // Save recommendation
        val recommendation = Recommendation(
            customerId = customerId,
            riskCategory = request.riskCategory,
            riskScore = request.riskScore,
            customerProfile = objectMapper.writeValueAsString(request.customerProfile)
        )
        recommendationRepository.save(recommendation)

        productsWithRationale.forEachIndexed { idx, product ->
            recommendedProductRepository.save(RecommendedProduct(
                recommendationId = recommendation.id,
                productId = product.productId,
                productName = product.name,
                category = product.category,
                matchScore = product.matchScore,
                allocationPercent = product.allocationPercent,
                rationale = product.rationale,
                rank = idx + 1
            ))
        }

        // Emit event
        kafkaTemplate.send(
            KafkaTopics.RECOMMENDATION_EVENTS,
            customerId.toString(),
            RecommendationGeneratedEvent(
                customerId = customerId,
                recommendationId = recommendation.id,
                riskCategory = request.riskCategory
            )
        )

        return RecommendationResponse(
            recommendationId = recommendation.id,
            customerId = customerId,
            riskCategory = request.riskCategory.name,
            products = productsWithRationale,
            generatedAt = recommendation.createdAt.toString()
        )
    }

    fun getLatestRecommendation(customerId: UUID): RecommendationResponse? {
        val rec = recommendationRepository.findFirstByCustomerIdOrderByCreatedAtDesc(customerId) ?: return null
        val products = recommendedProductRepository.findByRecommendationIdOrderByRankAsc(rec.id)

        return RecommendationResponse(
            recommendationId = rec.id,
            customerId = customerId,
            riskCategory = rec.riskCategory.name,
            products = products.map {
                RecommendedProductDto(
                    productId = it.productId, name = it.productName, category = it.category,
                    matchScore = it.matchScore, allocationPercent = it.allocationPercent,
                    rationale = it.rationale
                )
            },
            generatedAt = rec.createdAt.toString()
        )
    }

    private fun matchProducts(riskCategory: RiskCategory, profile: CustomerProfileDto): List<RecommendedProductDto> {
        // Rule-based product matching with scoring
        val products = getProductCatalog()

        return products
            .filter { isProductSuitable(it, riskCategory, profile) }
            .map { product ->
                val score = calculateMatchScore(product, riskCategory, profile)
                RecommendedProductDto(
                    productId = product.id,
                    name = product.name,
                    category = product.category,
                    matchScore = score,
                    allocationPercent = product.defaultAllocation.getOrDefault(riskCategory.name, 0.0)
                )
            }
            .sortedByDescending { it.matchScore }
            .take(8)
    }

    private fun reRankWithAI(products: List<RecommendedProductDto>, profile: CustomerProfileDto): List<RecommendedProductDto> {
        val productList = products.mapIndexed { idx, p ->
            "${idx + 1}. ${p.name} (${p.category}, match: ${p.matchScore}%)"
        }.joinToString("\n")

        val prompt = """Given this customer profile:
- Age group: ${profile.ageGroup}, Income: ${profile.incomeRange}
- Savings: ${profile.currentSavings}, Goals: retire with ${profile.retirementTarget}
- Risk tolerance: ${profile.riskCategory}

Re-rank these products from most to least suitable. Return ONLY the numbers in order, comma-separated:
$productList"""

        val response = chatModel.generate(prompt)
        val indices = response.trim().split(",").mapNotNull { it.trim().toIntOrNull()?.minus(1) }

        return if (indices.size == products.size) {
            indices.mapNotNull { products.getOrNull(it) }
        } else {
            products // fallback to original order
        }
    }

    private fun generateProductRationale(product: RecommendedProductDto, profile: CustomerProfileDto): String {
        val prompt = """In one sentence, explain why "${product.name}" (${product.category}) is suitable for a 
${profile.ageGroup} year old with ${profile.incomeRange} income and ${profile.riskCategory} risk profile 
who wants to retire with ${profile.retirementTarget}."""

        return chatModel.generate(prompt).trim()
    }

    private fun isProductSuitable(product: ProductInfo, riskCategory: RiskCategory, profile: CustomerProfileDto): Boolean {
        // Check risk level compatibility
        val riskCompatible = when (riskCategory) {
            RiskCategory.CONSERVATIVE -> product.riskLevel in listOf("LOW", "MODERATE")
            RiskCategory.MODERATE -> product.riskLevel in listOf("LOW", "MODERATE", "HIGH")
            RiskCategory.AGGRESSIVE -> product.riskLevel in listOf("MODERATE", "HIGH", "VERY_HIGH")
            RiskCategory.VERY_AGGRESSIVE -> product.riskLevel in listOf("HIGH", "VERY_HIGH")
        }
        return riskCompatible
    }

    private fun calculateMatchScore(product: ProductInfo, riskCategory: RiskCategory, profile: CustomerProfileDto): Double {
        var score = 50.0

        // Risk alignment
        score += when {
            product.riskLevel == riskCategory.name -> 25.0
            else -> 10.0
        }

        // Tax benefit bonus
        if (product.taxBenefit != null) score += 10.0

        // Tenure alignment with retirement horizon
        if (product.tenureYears != null && profile.retirementAge != null) {
            val yearsToRetire = profile.retirementAge - (profile.ageGroup?.substringBefore("-")?.toIntOrNull() ?: 30)
            if (product.tenureYears <= yearsToRetire) score += 15.0
        }

        return score.coerceIn(0.0, 100.0)
    }

    // In production, this would call the Product Catalog Service
    private fun getProductCatalog(): List<ProductInfo> = listOf(
        ProductInfo(UUID.randomUUID(), "High-Yield Savings Account", "SAVINGS_ACCOUNT", "LOW", null, null, mapOf("CONSERVATIVE" to 20.0, "MODERATE" to 15.0)),
        ProductInfo(UUID.randomUUID(), "Fixed Deposit - 5 Year", "FIXED_DEPOSIT", "LOW", "80C", 5, mapOf("CONSERVATIVE" to 30.0, "MODERATE" to 15.0)),
        ProductInfo(UUID.randomUUID(), "Large Cap Equity Fund", "MUTUAL_FUND", "MODERATE", null, null, mapOf("MODERATE" to 20.0, "AGGRESSIVE" to 15.0)),
        ProductInfo(UUID.randomUUID(), "Multi Cap Growth Fund", "MUTUAL_FUND", "HIGH", null, null, mapOf("AGGRESSIVE" to 25.0, "VERY_AGGRESSIVE" to 20.0)),
        ProductInfo(UUID.randomUUID(), "ELSS Tax Saver Fund", "TAX_SAVER", "MODERATE", "80C", 3, mapOf("MODERATE" to 10.0, "AGGRESSIVE" to 10.0)),
        ProductInfo(UUID.randomUUID(), "National Pension System", "PENSION", "LOW", "80CCD", null, mapOf("CONSERVATIVE" to 15.0, "MODERATE" to 10.0)),
        ProductInfo(UUID.randomUUID(), "Sovereign Gold Bond", "GOLD", "LOW", null, 8, mapOf("CONSERVATIVE" to 15.0, "MODERATE" to 10.0)),
        ProductInfo(UUID.randomUUID(), "PPF Account", "PPF", "LOW", "80C", 15, mapOf("CONSERVATIVE" to 20.0, "MODERATE" to 10.0)),
        ProductInfo(UUID.randomUUID(), "Small Cap Equity Fund", "MUTUAL_FUND", "VERY_HIGH", null, null, mapOf("AGGRESSIVE" to 15.0, "VERY_AGGRESSIVE" to 25.0)),
        ProductInfo(UUID.randomUUID(), "REITs Fund", "BOND", "HIGH", null, null, mapOf("AGGRESSIVE" to 10.0, "VERY_AGGRESSIVE" to 10.0)),
        ProductInfo(UUID.randomUUID(), "Corporate Bond Fund", "BOND", "MODERATE", null, 3, mapOf("CONSERVATIVE" to 10.0, "MODERATE" to 15.0)),
        ProductInfo(UUID.randomUUID(), "Term Insurance Plan", "INSURANCE", "LOW", "80C", null, mapOf("CONSERVATIVE" to 5.0, "MODERATE" to 5.0, "AGGRESSIVE" to 5.0))
    )
}

data class ProductInfo(
    val id: UUID,
    val name: String,
    val category: String,
    val riskLevel: String,
    val taxBenefit: String?,
    val tenureYears: Int?,
    val defaultAllocation: Map<String, Double>
)
